src/_pytest/hookspec.py

```diff
@@ -535,6 +545,27 @@ def pytest_logwarning(message, code, nodeid, fslocation):
     """
 
 
+@hookspec(historic=True)
+def pytest_warning_captured(warning_message, when, item):
+    """
+    Process a warning captured by the internal pytest warnings plugin.
+
+    :param warnings.WarningMessage warning_message:
+        The captured warning. This is the same object produced by :py:func:`warnings.catch_warnings`, and contains
+        the same attributes as the parameters of :py:func:`warnings.showwarning`.
+
+    :param str when:
+        Indicates when the warning was captured. Possible values:
+
+        * ``"config"``: during pytest configuration/initialization stage.
+        * ``"collect"``: during test collection.
+        * ``"runtest"``: during test execution.
```

Comment:
for better tracking i propose having not just runtest, but also the item phases called setup, call, teardown

Reply:
Oops just discovered that trying to catch warnings in `pytest_runtest_setup`, `pytest_runtest_call` and `pytest_runtest_teardown` cause problems with `recwarn` fixture. I don't see a way to fix this properly.

Any suggestions here @RonnyPfannschmidt ?

Reply:
OK, I'm reverting this to "runtest" then. 

If we implement support for `setup`, `call` and `teardown` in the future, we will have a breaking change in our hands because of `runtest`...

Reply:
actually, can we fake it by just implementing it like runtest, but switching the internal string based on phase?

Reply:
But warnings are cached by `warnings.catch_warnings`, so it seems out of our control.

Feel free to try out an idea and pushing to my branch if you have a clear picture on how to implement it. 👍